### PR TITLE
feat(`anvil`): add support for EIP4844 types

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

This does not include any functionality regarding 4844 yet other than support for 4844 variants in the corresponding types.

First of a series of PRs that will add Cancun support to foundry. Tracking here: https://github.com/foundry-rs/foundry/issues/5574

## Solution

Add support for 4844 transaction types.

### CODE REVIEW

File: crates/anvil/src/eth/fees.rs [hunk 1]
```diff
@@ -259,6 +259,12 @@ impl FeeHistoryService {
                                 .min(U256::from(t.max_fee_per_gas).saturating_sub(base_fee))
                                 .to::<u64>()
                         }
+                        // TODO: This probably needs to be extended to extract 4844 info.
```
Comment (Evalir) [inline crates/anvil/src/eth/fees.rs:262]: Will be done in a subsequent PR which modifies `FeeHistory` to accomodate 4844 changes

File: crates/anvil/core/src/eth/transaction/mod.rs [hunk 2]
```diff
@@ -132,6 +156,7 @@ pub enum TypedTransactionRequest {
     Legacy(TxLegacy),
     EIP2930(TxEip2930),
     EIP1559(TxEip1559),
+    EIP4844(TxEip4844),
```
Comment (mattsse) [inline crates/anvil/core/src/eth/transaction/mod.rs:159]: This is what we decode in send_rawTransaction right?
This needs to include the sidecar eventually.

for this

https://github.com/foundry-rs/foundry/blob/master/crates/anvil/src/eth/api.rs#L942-L942
we need another helper type that supports blobs rlp format

**Review by mattsse — CHANGES_REQUESTED** [review]
this does not support the envoloped format for 4844 which includes the sidecar

I think we want a helper type here that decodes with the sidecar, then extract the sidecar (+ validate), and convert into TypedTx

https://github.com/foundry-rs/foundry/blob/master/crates/anvil/src/eth/api.rs#L942-L942

File: crates/anvil/src/eth/fees.rs (hunk 1 above)
Comment (onbjerg) [reply crates/anvil/src/eth/fees.rs:262]: we should add a follow up issue for this to track

**Review by onbjerg — APPROVED** [review]
looks good to me, want @mattsse to take a look

File: crates/anvil/src/eth/fees.rs (hunk 1 above)
Comment (Evalir) [reply crates/anvil/src/eth/fees.rs:262]: This should be go on #7242

**Review by mattsse — APPROVED** [review]
lgtm

File: crates/anvil/core/src/eth/transaction/mod.rs [hunk 3]
```diff
@@ -39,11 +42,14 @@ pub fn transaction_request_to_typed(tx: TransactionRequest) -> Option<TypedTrans
         gas_price,
         max_fee_per_gas,
         max_priority_fee_per_gas,
+        max_fee_per_blob_gas,
```
Comment (gakonst) [inline crates/anvil/core/src/eth/transaction/mod.rs:45]: I forget, why don't we do this with a `From` impl / reuse any alloy types code? Is it due to the Deposit tx I guess?

Comment (gakonst) [reply crates/anvil/core/src/eth/transaction/mod.rs:45]: nit there's a `Some(126)` below i'd like to replace it with a const that indicates the semantic value of the number from op specs

File: crates/anvil/core/src/eth/transaction/mod.rs [hunk 4]
```diff
@@ -509,6 +618,8 @@ pub enum TypedTransaction {
     EIP2930(Signed<TxEip2930>),
     /// EIP-1559 transaction
     EIP1559(Signed<TxEip1559>),
+    /// EIP-4844 transaction
+    EIP4844(Signed<TxEip4844Variant>),
     /// op-stack deposit transaction
     Deposit(DepositTransaction),
```
Comment (gakonst) [inline crates/anvil/core/src/eth/transaction/mod.rs:624]: same here i take that we redefine the enum vs reusing from alloy bc of the deposit tx right?

@mattsse do u have any ideas for doing this with a super-enum?

something like `enum TypedTransaction { #[flatten] Inner(alloy::TypedTransaction), Deposit(...)`?

File: crates/anvil/src/eth/fees.rs (hunk 1 above)
Comment (gakonst) [reply crates/anvil/src/eth/fees.rs:262]: saw it there

**Review by gakonst — APPROVED** [review]
good w me @mattsse @Evalir lmk wyt on the point about the super enum

**Review by mattsse — APPROVED** [review]
lgtm, but this will likely break with new alloy rlp API

but since this most likely also breaks existing variants we can do this

**Pull request merged**